lib/src/main/java/com/knuddels/jtokkit/LazyEncodingRegistry.java

```diff
@@ -0,0 +1,26 @@
+package com.knuddels.jtokkit;
+
+import com.knuddels.jtokkit.api.Encoding;
+import com.knuddels.jtokkit.api.EncodingRegistry;
+import com.knuddels.jtokkit.api.EncodingType;
+
+import java.util.Optional;
+
+/**
+ * A lazy initialization implementation of {@link EncodingRegistry}. It does not register any encoding until either the 
+ * {@link #getEncoding(EncodingType)} or {@link #getEncoding(String)} method is called. 
+ * When one of these methods is called, the requested {@link EncodingType} is registered.
+ */
+public class LazyEncodingRegistry extends AbstractEncodingRegistry {
+    @Override
+    public Encoding getEncoding(EncodingType encodingType) {
+        addEncoding(encodingType);
```

Comment:
I believe, that this is not the behaviour you wanted in your issue: Now on every getEncoding call, the encoding file will be loaded from disk, regardless of whether it has already been loaded. We want to guarantee that it is only loaded once. Further more, I would like the guarantee to be, that even if multiple threads request the same encoding concurrently, it will only trigger a single file load.

I think we would benefit from double-checked locking here:
* Check if an encoding already exists for the encodingType (if yes, simply return it)
* If not, acquire a lock for loading that encoding (for locks, I would prefer if we not leak the lock)
* Check if an encoding exists after acquiring the lock (if yes, simply return it)
* If not, load the encoding, put it in the map and return it

Reply:
Yes, you're right. I didn't think it through thoroughly enough. Since encodings is a `ConcurrentHashMap`, I believe we can achieve the same effect without locking by using the `putIfAbsent` method in `addEncoding` of `AbstractEncodingRegistry`. Here is an example implementation:
```
 protected void addEncoding(EncodingType encodingType) {
        switch (encodingType) {
            case R50K_BASE:
                encodings.putIfAbsent(encodingType.getName(), EncodingFactory.r50kBase());
                break;
            case P50K_BASE:
                encodings.putIfAbsent(encodingType.getName(), EncodingFactory.p50kBase());
                break;
            case P50K_EDIT:
                encodings.putIfAbsent(encodingType.getName(), EncodingFactory.p50kEdit());
                break;
            case CL100K_BASE:
                encodings.putIfAbsent(encodingType.getName(), EncodingFactory.cl100kBase());
                break;
            default:
                throw new IllegalStateException("Unknown encoding type " + encodingType.getName());
        }
```
What do you think?

Reply:
You're completely right, that is a lot nicer than my proposal :smile: